In [34]:
import numpy as np
import pandas as pd

In [2]:
df1 = pd.read_csv("..\\Teoria\\titanic_1.csv")
df2 = pd.read_csv("..\\Teoria\\titanic_2.csv")

**1. Mostrar las estadisticas de cada uno de los dataframe**


In [3]:
df1

PassengerId                                               Name     Sex  \
0              1                            Braund, Mr. Owen Harris    male   
1              2  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2              3                             Heikkinen, Miss. Laina  female   
3              4       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4              5                           Allen, Mr. William Henry    male   
..           ...                                                ...     ...   
886          887                              Montvila, Rev. Juozas    male   
887          888                       Graham, Miss. Margaret Edith  female   
888          889           Johnston, Miss. Catherine Helen "Carrie"  female   
889          890                              Behr, Mr. Karl Howell    male   
890          891                                Dooley, Mr. Patrick    male   

      Age  
0    22.0  
1    38.0  
2    26.0  
3    35.0  
4    35.0  
..    ...  
886  27.0  
887  19.0  
888   NaN  
889  26.0  
890  32.0  

[891 rows x 4 columns]

In [4]:
df2

PassengerId  Survived  Pclass            Ticket     Fare
0              1         0       3         A/5 21171   7.2500
1              2         1       1          PC 17599  71.2833
2              3         1       3  STON/O2. 3101282   7.9250
3              4         1       1            113803  53.1000
4              5         0       3            373450   8.0500
..           ...       ...     ...               ...      ...
886          887         0       2            211536  13.0000
887          888         1       1            112053  30.0000
888          889         0       3        W./C. 6607  23.4500
889          890         1       1            111369  30.0000
890          891         0       3            370376   7.7500

[891 rows x 5 columns]

**1. Unir los dataframe para obtener solo uno**

In [5]:

df = df1.merge(right=df2, on="PassengerId")
df

PassengerId                                               Name     Sex  \
0              1                            Braund, Mr. Owen Harris    male   
1              2  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2              3                             Heikkinen, Miss. Laina  female   
3              4       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4              5                           Allen, Mr. William Henry    male   
..           ...                                                ...     ...   
886          887                              Montvila, Rev. Juozas    male   
887          888                       Graham, Miss. Margaret Edith  female   
888          889           Johnston, Miss. Catherine Helen "Carrie"  female   
889          890                              Behr, Mr. Karl Howell    male   
890          891                                Dooley, Mr. Patrick    male   

      Age  Survived  Pclass            Ticket     Fare  
0    22.0         0       3         A/5 21171   7.2500  
1    38.0         1       1          PC 17599  71.2833  
2    26.0         1       3  STON/O2. 3101282   7.9250  
3    35.0         1       1            113803  53.1000  
4    35.0         0       3            373450   8.0500  
..    ...       ...     ...               ...      ...  
886  27.0         0       2            211536  13.0000  
887  19.0         1       1            112053  30.0000  
888   NaN         0       3        W./C. 6607  23.4500  
889  26.0         1       1            111369  30.0000  
890  32.0         0       3            370376   7.7500  

[891 rows x 8 columns]

**2. Imprimir las últimas 5 filas de las columnas de nombre y edad**

In [6]:
df[["Name", "Age"]].tail(5)

Name   Age
886                     Montvila, Rev. Juozas  27.0
887              Graham, Miss. Margaret Edith  19.0
888  Johnston, Miss. Catherine Helen "Carrie"   NaN
889                     Behr, Mr. Karl Howell  26.0
890                       Dooley, Mr. Patrick  32.0

**3. Imprimir las primeras 10 filas de las instancias que en sexo sean igual a femenino**

In [7]:
df[df.Sex=="female"].head(10)

PassengerId                                               Name     Sex  \
1             2  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2             3                             Heikkinen, Miss. Laina  female   
3             4       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
8             9  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female   
9            10                Nasser, Mrs. Nicholas (Adele Achem)  female   
10           11                    Sandstrom, Miss. Marguerite Rut  female   
11           12                           Bonnell, Miss. Elizabeth  female   
14           15               Vestrom, Miss. Hulda Amanda Adolfina  female   
15           16                   Hewlett, Mrs. (Mary D Kingcome)   female   
18           19  Vander Planke, Mrs. Julius (Emelia Maria Vande...  female   

     Age  Survived  Pclass            Ticket     Fare  
1   38.0         1       1          PC 17599  71.2833  
2   26.0         1       3  STON/O2. 3101282   7.9250  
3   35.0         1       1            113803  53.1000  
8   27.0         1       3            347742  11.1333  
9   14.0         1       2            237736  30.0708  
10   4.0         1       3           PP 9549  16.7000  
11  58.0         1       1            113783  26.5500  
14  14.0         0       3            350406   7.8542  
15  55.0         1       2            248706  16.0000  
18  31.0         0       3            345763  18.0000

**4. Imprimir las instancias que en la columna Fare sean menor a 33 y en la columna Pclass igual a 2**

In [8]:
df[(df.Fare<33) & (df.Pclass==2)]

PassengerId                                          Name     Sex   Age  \
9             10           Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   
15            16              Hewlett, Mrs. (Mary D Kingcome)   female  55.0   
17            18                  Williams, Mr. Charles Eugene    male   NaN   
20            21                          Fynney, Mr. Joseph J    male  35.0   
21            22                         Beesley, Mr. Lawrence    male  34.0   
..           ...                                           ...     ...   ...   
866          867                  Duran y More, Miss. Asuncion  female  27.0   
874          875         Abelson, Mrs. Samuel (Hannah Wizosky)  female  28.0   
880          881  Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0   
883          884                 Banfield, Mr. Frederick James    male  28.0   
886          887                         Montvila, Rev. Juozas    male  27.0   

     Survived  Pclass            Ticket     Fare  
9           1       2            237736  30.0708  
15          1       2            248706  16.0000  
17          1       2            244373  13.0000  
20          0       2            239865  26.0000  
21          1       2            248698  13.0000  
..        ...     ...               ...      ...  
866         1       2     SC/PARIS 2149  13.8583  
874         1       2         P/PP 3381  24.0000  
880         1       2            230433  26.0000  
883         0       2  C.A./SOTON 34068  10.5000  
886         0       2            211536  13.0000  

[163 rows x 8 columns]

**5. Del resultado anterior obten el número de hombres y el número de mujeres**

In [9]:
[df[(df.Fare<33) & (df.Pclass==2) & (df.Sex=="male")].shape[0] ,
 df[(df.Fare<33) & (df.Pclass==2) & (df.Sex=="female")].shape[0]]

[95, 68]

**6. Obten el costo total de los tickets de hombres y el costo total de los tickets de mujeres**

In [10]:
[round(df[df.Sex=="male"].Fare.sum(),2),
 round(df[df.Sex=="female"].Fare.sum(),2)]

[14727.29, 13966.66]

**7. Obten el costo total de los tickets por cada clase (1, 2 y 3) y el número de pasajeros de cada clase**

In [11]:
for i in df.Pclass.unique():
    costo =  round(df[df.Pclass==i].Fare.sum(),2)
    npas = df[df.Pclass==i].shape[0]
    print("Para la clase", i, "=", costo, "€ y", npas, "pasajeros." )

Para la clase 3 = 6714.7 € y 491 pasajeros.
Para la clase 1 = 18177.41 € y 216 pasajeros.
Para la clase 2 = 3801.84 € y 184 pasajeros.


**8. Crear una columna con el apellido de cada pasajero y agregala al dataframe**

In [12]:
df_bu = df.copy()

In [13]:
df["Subname"] = df.Name.apply(lambda x: x.split(",")[0])
df

PassengerId                                               Name     Sex  \
0              1                            Braund, Mr. Owen Harris    male   
1              2  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2              3                             Heikkinen, Miss. Laina  female   
3              4       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4              5                           Allen, Mr. William Henry    male   
..           ...                                                ...     ...   
886          887                              Montvila, Rev. Juozas    male   
887          888                       Graham, Miss. Margaret Edith  female   
888          889           Johnston, Miss. Catherine Helen "Carrie"  female   
889          890                              Behr, Mr. Karl Howell    male   
890          891                                Dooley, Mr. Patrick    male   

      Age  Survived  Pclass            Ticket     Fare    Subname  
0    22.0         0       3         A/5 21171   7.2500     Braund  
1    38.0         1       1          PC 17599  71.2833    Cumings  
2    26.0         1       3  STON/O2. 3101282   7.9250  Heikkinen  
3    35.0         1       1            113803  53.1000   Futrelle  
4    35.0         0       3            373450   8.0500      Allen  
..    ...       ...     ...               ...      ...        ...  
886  27.0         0       2            211536  13.0000   Montvila  
887  19.0         1       1            112053  30.0000     Graham  
888   NaN         0       3        W./C. 6607  23.4500   Johnston  
889  26.0         1       1            111369  30.0000       Behr  
890  32.0         0       3            370376   7.7500     Dooley  

[891 rows x 9 columns]

**9. Encuentra si se repiten apellidos en la columna que se creo en el ejercicio anterior**

In [14]:
serie = df.Subname.value_counts()
L = serie[serie>1].index.values
L

array(['Andersson', 'Sage', 'Panula', 'Skoog', 'Carter', 'Goodwin',
       'Johnson', 'Rice', 'Lefebre', 'Asplund', 'Harris', 'Fortune',
       'Baclini', 'Brown', 'Hart', 'Kelly', 'Harper', 'Smith',
       'Gustafsson', 'Palsson', 'Williams', 'Ford', 'Johansson', 'Davies',
       'Jussila', 'Flynn', 'Hickman', 'Thayer', 'West', 'Laroche',
       'Bourke', 'Boulos', 'Navratil', 'Richards', 'Elias', "O'Brien",
       'Van Impe', 'Olsen', 'Goldsmith', 'Graham', 'Moran', 'Newell',
       'Collyer', 'Vander Planke', 'Hoyt', 'Meyer', 'Allison', 'Taussig',
       'Hansen', 'Jensen', 'Andrews', 'Moor', 'Rosblom', 'Pears', 'Lobb',
       'Hakkarainen', 'Olsson', 'Beckwith', 'Abbott',
       'Penasco y Castellana', 'Carlsson', 'Barbara', 'Duff Gordon',
       'Webber', 'Baxter', 'Attalah', 'Mellinger', 'Beane', 'Zabour',
       'Crosby', 'Nakid', 'Coutts', 'Minahan', 'Calic', 'Herman', 'Dick',
       'Ali', 'Abelson', 'Lam', 'Saad', 'Strom', 'McCoy', 'Jacobsohn',
       'Ryerson', 'Wick', 'Bish

**10. Encuentra los apellidos que se repiten en la columna recien creada**

In [15]:
serie = df.Subname.value_counts()
L = serie[serie>1].index.values
L

array(['Andersson', 'Sage', 'Panula', 'Skoog', 'Carter', 'Goodwin',
       'Johnson', 'Rice', 'Lefebre', 'Asplund', 'Harris', 'Fortune',
       'Baclini', 'Brown', 'Hart', 'Kelly', 'Harper', 'Smith',
       'Gustafsson', 'Palsson', 'Williams', 'Ford', 'Johansson', 'Davies',
       'Jussila', 'Flynn', 'Hickman', 'Thayer', 'West', 'Laroche',
       'Bourke', 'Boulos', 'Navratil', 'Richards', 'Elias', "O'Brien",
       'Van Impe', 'Olsen', 'Goldsmith', 'Graham', 'Moran', 'Newell',
       'Collyer', 'Vander Planke', 'Hoyt', 'Meyer', 'Allison', 'Taussig',
       'Hansen', 'Jensen', 'Andrews', 'Moor', 'Rosblom', 'Pears', 'Lobb',
       'Hakkarainen', 'Olsson', 'Beckwith', 'Abbott',
       'Penasco y Castellana', 'Carlsson', 'Barbara', 'Duff Gordon',
       'Webber', 'Baxter', 'Attalah', 'Mellinger', 'Beane', 'Zabour',
       'Crosby', 'Nakid', 'Coutts', 'Minahan', 'Calic', 'Herman', 'Dick',
       'Ali', 'Abelson', 'Lam', 'Saad', 'Strom', 'McCoy', 'Jacobsohn',
       'Ryerson', 'Wick', 'Bish

**11. Imprime los primeros 10 pasajeros que tengan un PassengerID multiplo de 3**

In [16]:
df[df.PassengerId%3==0].head(10)

PassengerId                                               Name     Sex  \
2             3                             Heikkinen, Miss. Laina  female   
5             6                                   Moran, Mr. James    male   
8             9  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female   
11           12                           Bonnell, Miss. Elizabeth  female   
14           15               Vestrom, Miss. Hulda Amanda Adolfina  female   
17           18                       Williams, Mr. Charles Eugene    male   
20           21                               Fynney, Mr. Joseph J    male   
23           24                       Sloper, Mr. William Thompson    male   
26           27                            Emir, Mr. Farred Chehab    male   
29           30                                Todoroff, Mr. Lalio    male   

     Age  Survived  Pclass            Ticket     Fare    Subname  
2   26.0         1       3  STON/O2. 3101282   7.9250  Heikkinen  
5    NaN         0       3            330877   8.4583      Moran  
8   27.0         1       3            347742  11.1333    Johnson  
11  58.0         1       1            113783  26.5500    Bonnell  
14  14.0         0       3            350406   7.8542    Vestrom  
17   NaN         1       2            244373  13.0000   Williams  
20  35.0         0       2            239865  26.0000     Fynney  
23  28.0         1       1            113788  35.5000     Sloper  
26   NaN         0       3              2631   7.2250       Emir  
29   NaN         0       3            349216   7.8958   Todoroff

**12. ¿Cuales son los apellidos de los pasajeros que son mujeres y menores de 20 años?**

In [17]:
df.Subname[(df.Sex=="female") & (df.Age<20)]

9         Nasser
10     Sandstrom
14       Vestrom
22       McGowan
24       Palsson
         ...    
852       Boulos
853        Lines
855          Aks
875        Najib
887       Graham
Name: Subname, Length: 75, dtype: object

**12. ¿Cuales son los apellidos de los pasajeros que son hombres y mayor de 45 años?**

In [18]:
df.Subname[(df.Sex=="male") & (df.Age>45)]

6             McCarthy
33             Wheadon
54               Ostby
92             Chaffee
94               Coxon
            ...       
771             Jensen
789         Guggenheim
851           Svensson
857               Daly
873    Vander Cruyssen
Name: Subname, Length: 73, dtype: object

**13. Compara si coinciden los apellidos generados en los ejercicios anteriores**

In [19]:
df_fe20 = df[["Subname"]][(df.Sex=="female") & (df.Age<20)]
df_ma45 = df[["Subname"]][(df.Sex=="male") & (df.Age>45)]

df_fe20.merge(right=df_ma45, on="Subname")

Subname
0  Johnson
1   Carter
2  Taussig
3   Harper

**14. Encuentra los 10 pasajeros que no sobrevivieron más jovenes e imprime cuantos son hombres y cuantos son mujeres**

In [20]:
df[(df.Survived==0)].sort_values(by="Age").head(10).groupby("Sex").size()

Sex
female    5
male      5
dtype: int64

**15. En un dataframe muestra 3 pasajeros que sean hombres y 3 que sean mujeres, que hayan sobrevivido y para cada clase de la columna Pclass**

In [21]:
filtrado = ["Pclass","Sex"]
df[(df.Survived==1)].groupby(by=filtrado).head(3).sort_values(by=filtrado)

PassengerId                                               Name     Sex  \
1             2  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
3             4       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
11           12                           Bonnell, Miss. Elizabeth  female   
23           24                       Sloper, Mr. William Thompson    male   
55           56                                  Woolner, Mr. Hugh    male   
97           98                    Greenfield, Mr. William Bertram    male   
9            10                Nasser, Mrs. Nicholas (Adele Achem)  female   
15           16                   Hewlett, Mrs. (Mary D Kingcome)   female   
43           44           Laroche, Miss. Simonne Marie Anne Andree  female   
17           18                       Williams, Mr. Charles Eugene    male   
21           22                              Beesley, Mr. Lawrence    male   
78           79                      Caldwell, Master. Alden Gates    male   
2             3                             Heikkinen, Miss. Laina  female   
8             9  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female   
10           11                    Sandstrom, Miss. Marguerite Rut  female   
36           37                                   Mamee, Mr. Hanna    male   
65           66                           Moubarek, Master. Gerios    male   
74           75                                      Bing, Mr. Lee    male   

      Age  Survived  Pclass            Ticket     Fare     Subname  
1   38.00         1       1          PC 17599  71.2833     Cumings  
3   35.00         1       1            113803  53.1000    Futrelle  
11  58.00         1       1            113783  26.5500     Bonnell  
23  28.00         1       1            113788  35.5000      Sloper  
55    NaN         1       1             19947  35.5000     Woolner  
97  23.00         1       1          PC 17759  63.3583  Greenfield  
9   14.00         1       2            237736  30.0708      Nasser  
15  55.00         1       2            248706  16.0000     Hewlett  
43   3.00         1       2     SC/Paris 2123  41.5792     Laroche  
17    NaN         1       2            244373  13.0000    Williams  
21  34.00         1       2            248698  13.0000     Beesley  
78   0.83         1       2            248738  29.0000    Caldwell  
2   26.00         1       3  STON/O2. 3101282   7.9250   Heikkinen  
8   27.00         1       3            347742  11.1333     Johnson  
10   4.00         1       3           PP 9549  16.7000   Sandstrom  
36    NaN         1       3              2677   7.2292       Mamee  
65    NaN         1       3              2661  15.2458    Moubarek  
74  32.00         1       3              1601  56.4958        Bing

**16. Obtener el tipo de elementos de cada columna y convierte los que sean int64 a int32. Evalua la memoria que consume antes y después de convertir las columnas y comenta la diferencia**

<class 'pandas.core.frame.DataFrame'>

RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Name         891 non-null    object 
 2   Sex          891 non-null    object 
 3   Age          714 non-null    float64
 4   Survived     891 non-null    int64  
 5   Pclass       891 non-null    int64  
 6   Ticket       891 non-null    object 
 7   Fare         891 non-null    float64
 8   Subname      891 non-null    object 
dtypes: float64(2), int64(3), object(4)
memory usage: 62.8+ KB

In [36]:
df_int32 = df.copy()

df_int32 = df_int32.astype({col: np.int32 for col in df_int32.select_dtypes(include=[np.int64]).columns})
df_int32.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int32  
 1   Name         891 non-null    object 
 2   Sex          891 non-null    object 
 3   Age          714 non-null    float64
 4   Survived     891 non-null    int32  
 5   Pclass       891 non-null    int32  
 6   Ticket       891 non-null    object 
 7   Fare         891 non-null    float64
 8   Subname      891 non-null    object 
dtypes: float64(2), int32(3), object(4)
memory usage: 52.3+ KB


**17. Encuentra a los pasajeros de la clase 1 y 3 entre 20 y 25 años y comprueba si son más hombres o mujeres**

In [75]:
con1 = ((df.Pclass==1)|(df.Pclass==3))
con2 = ((df.Age>20)|(df.Age<25))
df_17 = df[con1 & con2]
(df_17.Sex.value_counts().idxmax())

'male'

**18. Agrega una nueva columna al daframe que se llame: "Rango edad" y va a catagorizar de la siguiente manera:**
- Niño < 11 años
- Joven <18 años
- Adulto < 50 años
- Mayor >= 50 años

In [64]:
def rango(x):
    if x<11:
        return "Niño"
    if x<18:
        return "Joven"
    if x<50:
        return "Adulto"
    else:
        return "Mayor"

In [68]:
df["RangoEdad"]=df.Age.apply(rango)
df

PassengerId                                               Name     Sex  \
0              1                            Braund, Mr. Owen Harris    male   
1              2  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2              3                             Heikkinen, Miss. Laina  female   
3              4       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4              5                           Allen, Mr. William Henry    male   
..           ...                                                ...     ...   
886          887                              Montvila, Rev. Juozas    male   
887          888                       Graham, Miss. Margaret Edith  female   
888          889           Johnston, Miss. Catherine Helen "Carrie"  female   
889          890                              Behr, Mr. Karl Howell    male   
890          891                                Dooley, Mr. Patrick    male   

      Age  Survived  Pclass            Ticket     Fare    Subname RangoEdad  
0    22.0         0       3         A/5 21171   7.2500     Braund    Adulto  
1    38.0         1       1          PC 17599  71.2833    Cumings    Adulto  
2    26.0         1       3  STON/O2. 3101282   7.9250  Heikkinen    Adulto  
3    35.0         1       1            113803  53.1000   Futrelle    Adulto  
4    35.0         0       3            373450   8.0500      Allen    Adulto  
..    ...       ...     ...               ...      ...        ...       ...  
886  27.0         0       2            211536  13.0000   Montvila    Adulto  
887  19.0         1       1            112053  30.0000     Graham    Adulto  
888   NaN         0       3        W./C. 6607  23.4500   Johnston     Mayor  
889  26.0         1       1            111369  30.0000       Behr    Adulto  
890  32.0         0       3            370376   7.7500     Dooley    Adulto  

[891 rows x 10 columns]

**19. Encontrar a los pasajeros jovenes y mayores de la clase 2 y 3 que hayan sobrevivido. Imprimir los 10 primeros**

In [70]:
con1 = ((df.Pclass==2)|(df.Pclass==3))
con2 = ((df.RangoEdad=="Joven")|(df.RangoEdad=="Mayor"))
df_19 = df[con1 & con2]
df_19[df_19.Survived==1].head(10)

PassengerId                                 Name     Sex   Age  Survived  \
9            10  Nasser, Mrs. Nicholas (Adele Achem)  female  14.0         1   
15           16     Hewlett, Mrs. (Mary D Kingcome)   female  55.0         1   
17           18         Williams, Mr. Charles Eugene    male   NaN         1   
19           20              Masselmani, Mrs. Fatima  female   NaN         1   
22           23          McGowan, Miss. Anna "Annie"  female  15.0         1   
28           29        O'Dwyer, Miss. Ellen "Nellie"  female   NaN         1   
32           33             Glynn, Miss. Mary Agatha  female   NaN         1   
36           37                     Mamee, Mr. Hanna    male   NaN         1   
39           40          Nicola-Yarred, Miss. Jamila  female  14.0         1   
47           48            O'Driscoll, Miss. Bridget  female   NaN         1   

    Pclass  Ticket     Fare        Subname RangoEdad  
9        2  237736  30.0708         Nasser     Joven  
15       2  248706  16.0000        Hewlett     Mayor  
17       2  244373  13.0000       Williams     Mayor  
19       3    2649   7.2250     Masselmani     Mayor  
22       3  330923   8.0292        McGowan     Joven  
28       3  330959   7.8792        O'Dwyer     Mayor  
32       3  335677   7.7500          Glynn     Mayor  
36       3    2677   7.2292          Mamee     Mayor  
39       3    2651  11.2417  Nicola-Yarred     Joven  
47       3   14311   7.7500     O'Driscoll     Mayor

**20. Encuentra la edad promedio de todos los pasajeros y reemmplazala por los NaN en la columna de Age y ejecuta el ejercicio anterior**
- Nota: para identificar un nan utilizaremos np.nan

In [94]:
edadmedia = df.Age.mean(skipna=True)
df.Age.replace(np.nan, int(edadmedia),  inplace=True)
df

PassengerId                                               Name     Sex  \
0              1                            Braund, Mr. Owen Harris    male   
1              2  Cumings, Mrs. John Bradley (Florence Briggs Th...  female   
2              3                             Heikkinen, Miss. Laina  female   
3              4       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female   
4              5                           Allen, Mr. William Henry    male   
..           ...                                                ...     ...   
886          887                              Montvila, Rev. Juozas    male   
887          888                       Graham, Miss. Margaret Edith  female   
888          889           Johnston, Miss. Catherine Helen "Carrie"  female   
889          890                              Behr, Mr. Karl Howell    male   
890          891                                Dooley, Mr. Patrick    male   

      Age  Survived  Pclass            Ticket     Fare    Subname RangoEdad  
0    22.0         0       3         A/5 21171   7.2500     Braund    Adulto  
1    38.0         1       1          PC 17599  71.2833    Cumings    Adulto  
2    26.0         1       3  STON/O2. 3101282   7.9250  Heikkinen    Adulto  
3    35.0         1       1            113803  53.1000   Futrelle    Adulto  
4    35.0         0       3            373450   8.0500      Allen    Adulto  
..    ...       ...     ...               ...      ...        ...       ...  
886  27.0         0       2            211536  13.0000   Montvila    Adulto  
887  19.0         1       1            112053  30.0000     Graham    Adulto  
888  29.0         0       3        W./C. 6607  23.4500   Johnston     Mayor  
889  26.0         1       1            111369  30.0000       Behr    Adulto  
890  32.0         0       3            370376   7.7500     Dooley    Adulto  

[891 rows x 10 columns]

**21. Cambia a español la columna Sex**

In [102]:
df.Sex.replace(["male","female"], ["Hombre", "Mujer"], inplace=True)
df.rename(columns={'Sex': 'Genero'}, inplace=True)
df

PassengerId                                               Name  Genero  \
0              1                            Braund, Mr. Owen Harris  Hombre   
1              2  Cumings, Mrs. John Bradley (Florence Briggs Th...   Mujer   
2              3                             Heikkinen, Miss. Laina   Mujer   
3              4       Futrelle, Mrs. Jacques Heath (Lily May Peel)   Mujer   
4              5                           Allen, Mr. William Henry  Hombre   
..           ...                                                ...     ...   
886          887                              Montvila, Rev. Juozas  Hombre   
887          888                       Graham, Miss. Margaret Edith   Mujer   
888          889           Johnston, Miss. Catherine Helen "Carrie"   Mujer   
889          890                              Behr, Mr. Karl Howell  Hombre   
890          891                                Dooley, Mr. Patrick  Hombre   

      Age  Survived  Pclass            Ticket     Fare    Subname RangoEdad  
0    22.0         0       3         A/5 21171   7.2500     Braund    Adulto  
1    38.0         1       1          PC 17599  71.2833    Cumings    Adulto  
2    26.0         1       3  STON/O2. 3101282   7.9250  Heikkinen    Adulto  
3    35.0         1       1            113803  53.1000   Futrelle    Adulto  
4    35.0         0       3            373450   8.0500      Allen    Adulto  
..    ...       ...     ...               ...      ...        ...       ...  
886  27.0         0       2            211536  13.0000   Montvila    Adulto  
887  19.0         1       1            112053  30.0000     Graham    Adulto  
888  29.0         0       3        W./C. 6607  23.4500   Johnston     Mayor  
889  26.0         1       1            111369  30.0000       Behr    Adulto  
890  32.0         0       3            370376   7.7500     Dooley    Adulto  

[891 rows x 10 columns]

**22. Convierte a mayusculas todas las columnas que los valores sean cadenas**

In [105]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Name         891 non-null    object 
 2   Genero       891 non-null    object 
 3   Age          891 non-null    float64
 4   Survived     891 non-null    int64  
 5   Pclass       891 non-null    int64  
 6   Ticket       891 non-null    object 
 7   Fare         891 non-null    float64
 8   Subname      891 non-null    object 
 9   RangoEdad    891 non-null    object 
dtypes: float64(2), int64(3), object(5)
memory usage: 69.7+ KB


In [132]:
df_cadenas = df.copy()
serie_obj = df_cadenas.dtypes
serie_obj=="object"
col_obj = serie_obj[serie_obj=="object"].index.values
df_cadenas[col_obj] = df_cadenas[col_obj].applymap(lambda x: x.upper())
df_cadenas.head(50)

C:\Users\garch\AppData\Local\Temp\ipykernel_10088\1070626661.py:5: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_cadenas[col_obj] = df_cadenas[col_obj].applymap(lambda x: x.upper())


PassengerId                                               Name  Genero  \
0             1                            BRAUND, MR. OWEN HARRIS  HOMBRE   
1             2  CUMINGS, MRS. JOHN BRADLEY (FLORENCE BRIGGS TH...   MUJER   
2             3                             HEIKKINEN, MISS. LAINA   MUJER   
3             4       FUTRELLE, MRS. JACQUES HEATH (LILY MAY PEEL)   MUJER   
4             5                           ALLEN, MR. WILLIAM HENRY  HOMBRE   
5             6                                   MORAN, MR. JAMES  HOMBRE   
6             7                            MCCARTHY, MR. TIMOTHY J  HOMBRE   
7             8                     PALSSON, MASTER. GOSTA LEONARD  HOMBRE   
8             9  JOHNSON, MRS. OSCAR W (ELISABETH VILHELMINA BERG)   MUJER   
9            10                NASSER, MRS. NICHOLAS (ADELE ACHEM)   MUJER   
10           11                    SANDSTROM, MISS. MARGUERITE RUT   MUJER   
11           12                           BONNELL, MISS. ELIZABETH   MUJER   
12           13                     SAUNDERCOCK, MR. WILLIAM HENRY  HOMBRE   
13           14                        ANDERSSON, MR. ANDERS JOHAN  HOMBRE   
14           15               VESTROM, MISS. HULDA AMANDA ADOLFINA   MUJER   
15           16                   HEWLETT, MRS. (MARY D KINGCOME)    MUJER   
16           17                               RICE, MASTER. EUGENE  HOMBRE   
17           18                       WILLIAMS, MR. CHARLES EUGENE  HOMBRE   
18           19  VANDER PLANKE, MRS. JULIUS (EMELIA MARIA VANDE...   MUJER   
19           20                            MASSELMANI, MRS. FATIMA   MUJER   
20           21                               FYNNEY, MR. JOSEPH J  HOMBRE   
21           22                              BEESLEY, MR. LAWRENCE  HOMBRE   
22           23                        MCGOWAN, MISS. ANNA "ANNIE"   MUJER   
23           24                       SLOPER, MR. WILLIAM THOMPSON  HOMBRE   
24           25                      PALSSON, MISS. TORBORG DANIRA   MUJER   
25           26  ASPLUND, MRS. CARL OSCAR (SELMA AUGUSTA EMILIA...   MUJER   
26           27                            EMIR, MR. FARRED CHEHAB  HOMBRE   
27           28                     FORTUNE, MR. CHARLES ALEXANDER  HOMBRE   
28           29                      O'DWYER, MISS. ELLEN "NELLIE"   MUJER   
29           30                                TODOROFF, MR. LALIO  HOMBRE   
30           31                           URUCHURTU, DON. MANUEL E  HOMBRE   
31           32     SPENCER, MRS. WILLIAM AUGUSTUS (MARIE EUGENIE)   MUJER   
32           33                           GLYNN, MISS. MARY AGATHA   MUJER   
33           34                              WHEADON, MR. EDWARD H  HOMBRE   
34           35                            MEYER, MR. EDGAR JOSEPH  HOMBRE   
35           36                     HOLVERSON, MR. ALEXANDER OSKAR  HOMBRE   
36           37                                   MAMEE, MR. HANNA  HOMBRE   
37           38                           CANN, MR. ERNEST CHARLES  HOMBRE   
38           39                 VANDER PLANKE, MISS. AUGUSTA MARIA   MUJER   
39           40                        NICOLA-YARRED, MISS. JAMILA   MUJER   
40           41     AHLIN, MRS. JOHAN (JOHANNA PERSDOTTER LARSSON)   MUJER   
41           42  TURPIN, MRS. WILLIAM JOHN ROBERT (DOROTHY ANN ...   MUJER   
42           43                                KRAEFF, MR. THEODOR  HOMBRE   
43           44           LAROCHE, MISS. SIMONNE MARIE ANNE ANDREE   MUJER   
44           45                      DEVANEY, MISS. MARGARET DELIA   MUJER   
45           46                           ROGERS, MR. WILLIAM JOHN  HOMBRE   
46           47                                  LENNON, MR. DENIS  HOMBRE   
47           48                          O'DRISCOLL, MISS. BRIDGET   MUJER   
48           49                                SAMAAN, MR. YOUSSEF  HOMBRE   
49           50      ARNOLD-FRANCHI, MRS. JOSEF (JOSEFINE FRANCHI)   MUJER   

     Age  Survived  Pclas

**23. Agrupa por sexo y rango de edad y encuentra la edad promedio**

In [130]:
df.groupby(["Genero","RangoEdad"]).Age.mean()

Genero  RangoEdad
Hombre  Adulto       30.196793
        Joven        15.320000
        Mayor        37.750000
        Niño          3.959697
Mujer   Adulto       30.260870
        Joven        15.229167
        Mayor        36.626667
        Niño          4.596774
Name: Age, dtype: float64

**24. Agrupa por sobrevivientes, sexo y encuentra la edad promedio**

In [131]:
df.groupby(["Genero","Survived"]).Age.mean()

Genero  Survived
Hombre  0           31.013889
        1           27.529083
Mujer   0           25.876543
        1           28.871245
Name: Age, dtype: float64

**25. Divide la columna name en las siguientes columnas:**
 - Mrs/Miss
 - Apellido
 - Nombre(s)
 
 
Nota: Ignora el resto de información


In [139]:
df_noms = df.copy()
df_noms["Mrs/Miss"] = df.Name.apply(lambda x: x.split(",")[1].split(".")[0])
df_noms["Apellido"] = df.Name.apply(lambda x: x.split(",")[0])
df_noms["Nombre"] = df.Name.apply(lambda x: x.split(",")[1].split(".")[1])
df_noms

PassengerId                                               Name  Genero  \
0              1                            Braund, Mr. Owen Harris  Hombre   
1              2  Cumings, Mrs. John Bradley (Florence Briggs Th...   Mujer   
2              3                             Heikkinen, Miss. Laina   Mujer   
3              4       Futrelle, Mrs. Jacques Heath (Lily May Peel)   Mujer   
4              5                           Allen, Mr. William Henry  Hombre   
..           ...                                                ...     ...   
886          887                              Montvila, Rev. Juozas  Hombre   
887          888                       Graham, Miss. Margaret Edith   Mujer   
888          889           Johnston, Miss. Catherine Helen "Carrie"   Mujer   
889          890                              Behr, Mr. Karl Howell  Hombre   
890          891                                Dooley, Mr. Patrick  Hombre   

      Age  Survived  Pclass            Ticket     Fare    Subname RangoEdad  \
0    22.0         0       3         A/5 21171   7.2500     Braund    Adulto   
1    38.0         1       1          PC 17599  71.2833    Cumings    Adulto   
2    26.0         1       3  STON/O2. 3101282   7.9250  Heikkinen    Adulto   
3    35.0         1       1            113803  53.1000   Futrelle    Adulto   
4    35.0         0       3            373450   8.0500      Allen    Adulto   
..    ...       ...     ...               ...      ...        ...       ...   
886  27.0         0       2            211536  13.0000   Montvila    Adulto   
887  19.0         1       1            112053  30.0000     Graham    Adulto   
888  29.0         0       3        W./C. 6607  23.4500   Johnston     Mayor   
889  26.0         1       1            111369  30.0000       Behr    Adulto   
890  32.0         0       3            370376   7.7500     Dooley    Adulto   

    Mrs/Miss   Apellido                                  Nombre  
0         Mr     Braund                             Owen Harris  
1        Mrs    Cumings   John Bradley (Florence Briggs Thayer)  
2       Miss  Heikkinen                                   Laina  
3        Mrs   Futrelle           Jacques Heath (Lily May Peel)  
4         Mr      Allen                           William Henry  
..       ...        ...                                     ...  
886      Rev   Montvila                                  Juozas  
887     Miss     Graham                          Margaret Edith  
888     Miss   Johnston                Catherine Helen "Carrie"  
889       Mr       Behr                             Karl Howell  
890       Mr     Dooley                                 Patrick  

[891 rows x 13 columns]